ci: merge staging to master

maxwell-aiden opened the pull request:
This is an automatic PR generated by the pipeline CI/CD. This will be automatically fast-forward merged if successful.

emmacasolin committed:
ci: test load balancing

Introduced sharding for macOS and Windows runners and test directory pipelines for Linux runner

emmacasolin committed:
ci: improved junit output

- Includes test suites that failed to run
- Adds separator between names of nested describes

emmacasolin committed:
ci: optimising Homebrew setup

Stopped Mac CI jobs from updating dependencies since this was increasing setup time

emmacasolin committed:
ci: chocolatey caching

Optimised Windows CI/CD setup by internalising and caching chocolatey packages

Fixes #397

tegefaulkes committed:
fix: `findNode` and `getClosestGlobalNodes` no longer throws `ErrorNodeGraphEmptyDatabase` with empty network

`getClosestGlobalNodes` was throwing `ErrorNodeGraphEmptyDatabase` when it failed to get new nodes during the search process. Now it just returns undefined as expected.

Related #398

tegefaulkes committed:
test: adding docker integration test

emmacasolin committed:
ci: `build:dist` is now before `build:platforms-generate`

emmacasolin committed:
fix: executable permissions for `choco-install.ps1`

emmacasolin committed:
ci: moving package installations into main script for mac/windows

emmacasolin committed:
fix: parameter fixes for chocolatey install script

emmacasolin committed:
ci: cache set to always (even if a job fails) for all paths

emmacasolin committed:
Merge pull request #394 from MatrixAI/feature-test-load-balancing

Test load balancing (migrating changes from TypeScript-Demo-Lib)

tegefaulkes committed:
feat: added bin utils functions for running target commands

This is to be used with the pkg integration tests to test the docker, windows and mac executables.

Related #410

tegefaulkes committed:
feat: added switching utility functions

These return either the pkX or the pkXTarget command if the `cmd` is set.

Related #410

tegefaulkes committed:
fix: updated `start.test.ts` to use #410 changes

tegefaulkes committed:
fix: updated `bootstrap.test.ts` to use #410 changes

tegefaulkes committed:
feat: adding tmdDir to globals

Allows us to override the tmp used for testing with `PK_TMP_DIR`.

CMCDragonkai committed:
npm: set `pkg` script to run with `node` explicitly to help with windows

tegefaulkes committed:
test: converting tests for docker target

- converting lock.test.ts
- cleaning up `testIf` commands and envs
- converted lockall.test.ts
- converted status.test.ts
- converted stop.test.ts
- converted unlock.test.ts

Related #407

emmacasolin committed:
build: upgrading NodeJS to 16.15.0

tegefaulkes committed:
feat: added ability to override keypair generation with `privateKeyOverride` parameter for `createKeyManager`

This will skip key generation and use the provided `PrivateKey` instead. This should speed up testing by skipping the key generation.

Related #404

tegefaulkes committed:
feat: added `--private-key-file` option to `CommandStart.ts` and `CommandBootstrap.ts`

This should allow us to override the keypair generation with the provided private key. this will speed up agent starting.

Note that the key is provided as a Pem. The `PrivateKey` type contains functions that get destroyed somewhere between `commandStart.ts` and `keyManager.createKeyManager`. So I'm using the Pem string to keep the type primitive

Added `PK_ROOT_KEY` ENV and changed `--private-key-file` to `--root-key-file`

Related #404

CMCDragonkai committed:
chore: added entry about `GITHUB_TOKEN` in `.env.example`

CMCDragonkai committed:
chore: added `.npmrc` to reify npm link configuration, and using `npm pkg get name` to get package name in `shell.nix`

CMCDragonkai committed:
chore: synchronised `.npmignore` with TypeScript-Demo-Lib

CMCDragonkai committed:
style: lint now allowed PascalCase and UPPER_CASE for enum members, and `// ts-ignore` now requires a comment to be allowed

tegefaulkes committed:
test: added `setupTestAgent` for creating test agents

This is a quick and easy way to create an agent with a pre-generated key.

Related #404

tegefaulkes committed:
test: removing global agent usage

I've also used the key override where needed for `createPolykeyAgent`.

Related #404

CMCDragonkai committed:
ci: swapped to using `scripts/brew-install.sh`

CMCDragonkai committed:
style: moved default below variables for `scripts/check-test-generate.sh` and added bash sanity checks

CMCDragonkai committed:
style: moved `CI_PARALLEL` to the top for `scripts/build-platforms-generate.sh`

CMCDragonkai committed:
style: lintfixed `scripts/pkg.js`

CMCDragonkai committed:
ci: only bundle prebuilds in `scripts/pkg.js`

CMCDragonkai committed:
nix: `npm run build` will use `npm_config_nodedir` as an environment variable to propagate to potential prebuild scripts

CMCDragonkai committed:
chore: lint and lintfix scripts apply to `scripts` directory too

CMCDragonkai committed:
feat: swapped `ts-node` to using `swc` and removed `typescript-cached-transpile`

CMCDragonkai committed:
ci: no need to install python as no native addons are being built

CMCDragonkai committed:
ci: no need to use `--runInBand` for linux tests

CMCDragonkai committed:
fix: typo in `.gitlab-ci.yml` about running in staging, should be running in feature branches

CMCDragonkai committed:
ci: inherit `$PARENT_PIPELINE_ID` for child pipelines to refer to in the future

tegefaulkes committed:
fix: edge case bug for discovery queue

it wasn't handling the `ErrorConnectionNotRunning` that occurred in a very specific case.

tegefaulkes committed:
test: converted tests

Related #407

tegefaulkes committed:
test: overriding keypair generation for most tests

This should allow for faster testing since we skip the CPU intensive keypair generation.

Related #420

tegefaulkes committed:
tests: fixing tests for docker target

Related #407

tegefaulkes committed:
chore: updating the `.env.example` file

Related #407

tegefaulkes committed:
fix: tests use `global.tmpDir` directory

Also enabled `FF_NETWORK_PER_BUILD` flag for CI.

Related #407

emmacasolin committed:
feat: CLI format option affects logger output

emmacasolin committed:
chore: synchronised ci changes with TypeScript-Demo-Lib

- Reverted Windows to NodeJS v16.14.2
- Added `--arg ci true` to `nix-shell`
- Introduced continuous benchmarking

CMCDragonkai committed:
docs: updated README.md with information from Polykey.old

CMCDragonkai committed:
chore: changed `js-polykey` to `Polykey` and using `polykey` as the npm name [ci skip]

tegefaulkes committed:
fix: adding `--rm` to `docker-run.sh` script.

This will remove the container when it ends.

#407

tegefaulkes committed:
tests: disabling tests

Disabling tests that don't work with docker for now.

#407

CMCDragonkai committed:
refactor: changed from IIFE to `void main();` [ci skip]

tegefaulkes committed:
ci: cleaning up CI jobs

tegefaulkes committed:
fix: removing `pkXSwitch` in favor of `pkX`

#410

tegefaulkes committed:
fix: moved `globalRootKeyPems.ts` to fixtures

#404
#420

tegefaulkes committed:
fix: removing `docker-run.sh` script

#407

tegefaulkes committed:
fix: bug with `swc` and decorators

Due to a difference with swc on how decorators are handled. Classes were constructed without decorators applied when called inside of `createX` static functions. This was solved by using `new this()` to construct instead.

Fixed up some linting errors as well.

CMCDragonkai committed:
Merge pull request #430 from MatrixAI/feature-swc_fix

Fixing `swc` specific bug with decorators and constructors

tegefaulkes committed:
feat: `pkXTarget` works with any shell command

`DOCKER_OPTIONS` env variable containing all the docker run parameters to make the test work is provided to the command.

#410

tegefaulkes committed:
fix: changing `PK_TEST_DATA_DIR` to `PK_TEST_TMPDIR`

tegefaulkes committed:
fix: `PK_TEST_COMMAND` and PK_TEST_PLATFORM` must be set at the same time

The `runTestIfPlatform` commands will default to running if `PK_TEST_PLATFORM` is not set.

#410

tegefaulkes committed:
fix: updating nat tests to use `runDescribeIf`

emmacasolin committed:
fix: ts-ignoring logger type incompatibilities

tegefaulkes committed:
fix: consolidated all spawning functions in `tests/utils/exec.ts`

This is to key every major spawning command in one place in case we need to modify how we spawn using `ts-node`.

tegefaulkes committed:
fix: escaping args when using `shell: true`

tegefaulkes committed:
fix: `pkExecTarget` now uses `child_process.spawn`

Is is due to a compatability problems when using `shell: true` with `child_process.execFile`.

CMCDragonkai committed:
Merge pull request #391 from MatrixAI/feature-pkg_integration_tests

Packaged executable integration tests - initial integration tests with docker

CMCDragonkai committed:
Merge pull request #421 from MatrixAI/feature-json-logger

CLI option `--format=json` should change the STDERR logging to output JSON structured logging

CMCDragonkai committed:
chore: disable windows and macos build stage and integration stage jobs

CMCDragonkai committed:
style: changed to lowercase `npm_config_*` env variables

CMCDragonkai committed:
ci: change to using `--arg ci true` for all jobs using `nix-shell`

CMCDragonkai committed:
ci: switch back to using 16.15.1 nodejs for windows jobs

CMCDragonkai committed:
ci: `integration:docker` should be using gitlab variables and fixed styling

CMCDragonkai committed:
chore: update `encryptedfs` to 3.5.5

CMCDragonkai committed:
style: disallow newlines between import groups

emmacasolin committed:
feat: process execution standardisation

Standardised process execution for bin tests using `pkExec` and `pkSpawn`.

- All usage of `pkExec`/`pkSpawn` calls underlying default `WithoutShell` or override `WithShell` methods
- Combined `env` and `cwd` options into an `ExecOpts` object and added `command` and `shell` options
- Refactored `pkSpawnNs` and `pkExecNs` to use standardised methods
- Replaced all usage of `exec`/`execFile` with `spawn`
- error handling for generic exec/spawn + using new generic spawn in NAT utils
- ignore coverage from `src/proto`
- removed spawnfile
- added a check for if spawning process doesn't properly start

tegefaulkes committed:
test: Moving `utils.ts` into `tests/utils`

Moved `tests/utils.ts` to `tests/utils/utils.ts` and created `test/utils/index.ts`. Trying to keep the utils within one place.

Removed `setupGlobalAgent` code, #420 means it isn't used anymore.

Updating how we handle conditional testing utility's. we're switching to using reified booleans and just composing conditional boolean expression.

#434

tegefaulkes committed:
test: creating reified booleans for conditionally enabling tests

Creating `isPlatformX`, `isTestPlatformX` and `hasX` booleans to use for enabling tests.

Replaced usage of describeIf to use testIf for the tests within the describe block. This means jest shouldn't complain about empty test files if no tests run.

#434

tegefaulkes committed:
test: removing usage of `runTestIfPlatforms` and `runDescribeIfPlatforms`, replaced with `testIf`

We're replacing these with `testIf` and `describeIf` and reified booleans.

#434

tegefaulkes committed:
fix: replacing `global` with `globalThis` and adding linting rule

We should be using `globalThis` to access global variables. A linting rule has been added to restrict usage of `global`.

tegefaulkes committed:
test: importing test utils as `testUtils`

CMCDragonkai committed:
Merge pull request #437 from MatrixAI/feature-conditional_testing_fix

Standardise conditional testing fix

tegefaulkes committed:
fix: vaults locking and transactional locking

emmacasolin committed:
fix: using `Entrypoint` instead of  `Cmd` in `release.nix`

tegefaulkes committed:
build: bringing in `fast-check`, `jest-extended` and updating `@matrixai/db`

tegefaulkes committed:
build: adding `benches` to linting targets

tegefaulkes committed:
fix: moving `PromiseDeconstructed` to types

tegefaulkes committed:
fix: updating `setupFilesAfterEnv`

Not sure what these changes do. I don't know what to name it.

tegefaulkes committed:
ci: disabling some check tests

CMCDragonkai committed:
Merge pull request #436 from MatrixAI/feature-process-exec

Standardise process execution for tests

CMCDragonkai committed:
ci: coverage reports should be ignoring `src/proto/**`

emmacasolin committed:
fix: removed `/bin/polykey` from `PK_TEST_COMMAND`

CMCDragonkai committed:
Merge pull request #439 from MatrixAI/feature-docker-exec

Updates to Docker execution

CMCDragonkai committed:
docs: updated AWS ECR manual deployment instructions

CMCDragonkai committed:
fix: default node path should be using `path.join` for cross platform path construction

tegefaulkes committed:
fix: cleaning up unnecessary locks

This includes using `getForUpdate` for any counter updates.

tegefaulkes committed:
fix: some get/set operations just use DB if transaction not given.

tegefaulkes commented on tests/utils/exec.ts:
```diff
@@ -0,0 +1,600 @@
+  shell?: boolean;
+};
+
+const tsConfigPath = path.resolve(
+  path.join(globalThis.projectDir ?? '', 'tsconfig.json'),
+);
+
+const polykeyPath = path.resolve(
+  path.join(globalThis.projectDir ?? '', 'src/bin/polykey.ts'),
+);
```
`globalThis.projectDir ?? ''` had to be added to prevent an error for `globalThis.projectDir` being undefined. Likely the same needs to be done for `globalThis.testDir` somewhere.

Details here https://github.com/MatrixAI/Polykey/pull/436#issuecomment-1210251745

CMCDragonkai replied:
Why would this be affected by https://github.com/MatrixAI/Polykey/pull/411#issuecomment-1212635793?

This problem was due to `ts-node` running an internal test script that did not have the test environment.

The docker tests I feel like has too much required environment variables, so we want to remove them as much as  possible. At least the `PK_TEST_PLATFORM` should disappear with #435.

tegefaulkes replied:
The problem was that `globalThis` variables are not defined when running as a sub process. That means it won't be defined when running as docker as well. If the docker image is trying to use `globalThis.testDir` somewhere then we will run into an error `globalThis.testDir` is undefined. As far as the type system knows  the `globalThis` variables are always defined so it's not caught when building.

CMCDragonkai replied:
What docker process would be using `globalThis`? It would only be possible if you were running a test script under `tests` dir that imports the `tests/utils`. The docker tests are not running test scripts are they? They should be running `src/bin/polykey.ts` which wouldn't be using `globalThis.testDir`...

tegefaulkes replied:
It shouldn't be using `globalThis`. But if this problem is anything like the previous problem where just importing `tests/utils/exec.ts` was causing `global.this` variables to be used then we would get this error. I'll have to dig into why this is happening but it's too similar to the previous problem to not check for that first.

CMCDragonkai replied:
Let's do a quickfix for this. But long term we need to move all the integration tests into a separate directory and discard these hacks. Instead we would do things like:

```
import parameterisedTest from '..somecommontestmodule..';

describe('', () => {
  test('', parameterisedTest(p1, p2));
});
```

The idea is that `parameterisedTest(p1, p2)` should return an `() => Promise<void>` function to execute as the actual test.

Doing this though, we need to be careful about any side-effects coming from `beforeEach` and `beforeAll` and `afterEach` and `afterAll`. In fact, it would make sense to ensure that all of our tests are functional, which means if things need to be established before a test, they musts be passed as a pointer into the test function itself, so they are still pointing to the same objects being mutated.

This way we separate the "test structure" from the actual test itself which can be shared among different test files.

tegefaulkes replied:
I've narrowed it down. the `--node-path` parameter we're giving in the test is not properly relative to the working directory. Leading us to try and create the agent at a path that doesn't exist. 

```
// docker
// node path
'tmp/test/polykey-test-NCPSFm/polykey'

// working directory
/home/faulkes/matrixCode/polykey/js-polykey/tmp/test/polykey-test-NCPSFm
```

```
// normally
// node path
'/tmp/polykey-test-WWjNME/polykey'

// working directory
/tmp/polykey-test-WWjNME
```

Ah I see, normally we're using a full absolute path, but with the docker testing we're using a relative path. To fix this we should be using consistent in the style of path for both cases. The simplest change would be to make dataDir an absolute path in all cases by using `path.resolve`.

Also the `testCmd` was not getting passed to the `pkSpawn` opts.

tegefaulkes replied:
There's a single test failure now. 

```
FAIL tests/bin/polykey.test.ts
  ● polykey › format option affects STDERR
    SyntaxError: Unexpected token d in JSON at position 0
        at JSON.parse (<anonymous>)
      63 |       rlErr.once('close', reject);
      64 |     });
    > 65 |     const stderrParsed = JSON.parse(stderrStart);
         |                               ^
      66 |     expect(stderrParsed).toMatchObject({
      67 |       level: expect.stringMatching(/INFO|WARN|ERROR|DEBUG/),
      68 |       key: expect.any(String),
      at Object.<anonymous> (tests/bin/polykey.test.ts:65:31)
      
```

This is odd, I'm not sure what's causing it and it's passing locally. I'm running the test again with logging enabled to try and sus it out. But if I can't see a quick solution I'll just disable the test.

tegefaulkes replied:
Found and fixed the problem. Resolving this.

tegefaulkes committed:
ci: reenabling check tests

tegefaulkes committed:
build: updating `@matrixai/db` to `^5.0.3` and `encryptedfs` to `^3.5.6`

tegefaulkes committed:
fix: removed `ts-ignore`s for logger version mismatch

tegefaulkes committed:
fix: using `tran.getForUpdate` for vaults renaming and sigchain claim creation

tegefaulkes committed:
fix: added transaction lock on vaultId for vault renaming

tegefaulkes committed:
fix: `VaultManager` `vaultId` locking path and using `join('')` on paths

tegefaulkes committed:
Merge pull request #419 from MatrixAI/feature-concurrent-state

Integrate js-db and concurrency testing

CMCDragonkai committed:
fix: integrating `Timer` with `PromiseCancellable` from `@matrixai/async-cancellable`

* added `isPromise`, `isPromiseLike`, `isIterable`, `isAsyncIterable` to detect async and generator interfaces
* timed decorator works for regular values, `PromiseLike` and `Iterable` and `AsyncIterable`
* introduced `ContextTimed` type and other `Context*` types
* stack trace is refers when construction time, so decorator takes error class constructor

CMCDragonkai committed:
fix: `utils.sleep` now returns `Promise<void>`

CMCDragonkai committed:
fix: `globalTeardown.ts` now forces the removal of `globalDataDir`

tegefaulkes committed:
fix: `Queue` using `EventTarget` for task promises

tegefaulkes committed:
fix(tasks): adding `Task` getter functions `getTask` `getTasks` and `getGroupTasks`

tegefaulkes committed:
fix(tasks): changing `TaskGroup` to `TaskPath` as a alias for `LevelPath`, `taskGroup` is refered to as `path` now

tegefaulkes committed:
fix(tasks): making `Task` type a POJO of certain task data and a promise

CMCDragonkai committed:
npm: added `ts-node-inspect` in order to start the debugging port

CMCDragonkai committed:
chore: updated `@types/node` and typescript

CMCDragonkai committed:
feat(contexts): `timed` and `cancellable` higher order function operators

CMCDragonkai committed:
fix: `Timer` is clipped to maximum timeout if given finite delay

CMCDragonkai committed:
fix(contexts): changed timeout exception to `ErrorContextsTimedTimeout`

CMCDragonkai committed:
feat: introducing `Timer` as an object for tracking `setTimeout`, to be used for async deadlines

CMCDragonkai committed:
feat: introducing `timed` and `cancellable` decorators to automate asynchronous deadlines and asynchronous cancellation

CMCDragonkai committed:
feat: introducing `tasks` domain for managing background asynchronous tasks

* Created a `Plug` class for managing the locking of scheduling and queuing loops
* Using `performance.now()` for acquiring the current time

CMCDragonkai committed:
build: target ES2022

CMCDragonkai committed:
fix(timer): Timer now maintains event loop reference, and can have infinite timers

CMCDragonkai committed:
fix(contexts): decorators should check for generators, not iterables

CMCDragonkai committed:
feat(tasks): centralising Queue and Scheduler into a single `TaskManager`

* tasks can be cancelled at any stage: scheduled, queued or active
* `TaskData` is suitable to be encoded into JSON and back
* Graceful shutdown of `TaskManager`
* `TaskHandler` gets `TaskInfo` as second parameter after the `ContextTimed`

CMCDragonkai committed:
style: updated eslint dependencies

CMCDragonkai committed:
style: allow throwing literals due to abort signal reasons

CMCDragonkai committed:
style: linting timer and contexts

tegefaulkes committed:
tests: fixing concurrency limit test

tegefaulkes committed:
style: linting `Tasks` tests

tegefaulkes committed:
tests: expanding encode/decode tests

tegefaulkes committed:
fix: set `TaskManager` test logger level to `WARN`

tegefaulkes committed:
tests: removing old `Queue` and `Scheduler` tests

tegefaulkes committed:
build(timer): removing `timer` domain and adding `@matrixai/timer` dependency

tegefaulkes committed:
feat: `PolykeyAgent.ts` using `TaskManager`

tegefaulkes committed:
feat: updated `NodeManager` to use `TaskManager`

CMCDragonkai committed:
fix(contexts): when timed decorator inherits timer and signal, it should do nothing

There are 3 properties for the `timed` wrapper:

A. If timer times out, signal is aborted
B. If signal is aborted, timer is cancelled
C. If timer is owned by the wrapper, then it must be cancelled when the
target finishes

There are 4 cases where the wrapper is used and where the properties are
applied:

  1. Nothing is inherited - A B C
  2. Signal is inherited - A B C
  3. Timer is inherited - A
  4. Both signal and timer are inherited - A*

B and C are only applied to case 1 and 2, because that's when the `Timer`
is owned by the wrapper.

*Case 4 is a special case, because the timer and signal are inherited,
so it is assumed that the handlers are already setup betwen the timer
and signal.

tegefaulkes committed:
feat: updated `NodeConnectionManager` to use `TaskManager`

CMCDragonkai committed:
feat(contexts): introducing `timedCancellable` decorator and HOF and factored out common functionality in contexts domain

tegefaulkes committed:
tests: cleaing up `TaskManager.test.ts`

tegefaulkes committed:
fix: fixing type bug in `Discovery`

tegefaulkes committed:
Merge pull request #438 from MatrixAI/feature-queue

Feature TaskManager Scheduler and Queue and Context Decorators for Timed and Cancellable

tegefaulkes committed:
feat: adding cancellability to `NodeManager` handlers

tegefaulkes committed:
fix: small bug with `pingNode`

tegefaulkes committed:
fix: bugs with `nodeConnectionManager.syncNodeGraph`

tegefaulkes committed:
tests: cleaning up dependencies in tests

tegefaulkes committed:
fix: `getRemoteNodeClosestNodes` shouldn't throw connection errors

`getRemoteNodeClosestNodes` was throwing an connection error in certain conditions. If it failed to connect to a node it should've just skipped that node.

#418

tegefaulkes committed:
fix: excessive connections from `refreshBuckets`

Removing excessive logging for using connections. We don't need a 3 log messages for each time we use an existing connection.

Adding 'jitter' or spacing to the `refreshBuckets` delays so that they don't run all at once. This is implemented with a `refreshBucketDelaySpread` paramater that specifies the multiple of the delay to spread across. defaults to 0.5 for 50%

Adding a 'heuristic' to `refreshBucket` to prevent it from contacting the same nodes repeatably. Currently this is just a check in `getClosestGlobalNodes` where if we find less than `nodeBucketLimit` nodes we just reset the timer on all `refreshBucket` tasks.

Adding tests for checking the spread of `refreshBucket` delays. Another test for resetting the timer on `refreshBucket` tasks if a `findNode` finds less than 20 nodes.

#415

tegefaulkes committed:
feat: `nodeConnectionManager.getClosestGlobalNodes` can optionally skip recently offline nodes

This is done with an in-memory map of `nodeIdstring` to some data tracking the backoff period. it defaults to 5 min and doubles each failure.

#413

tegefaulkes committed:
feat: added handler to detect promise deadlocks to `ExitHandlers.ts`

#307

tegefaulkes committed:
fix: handerIds are derived from class and handler function names

tegefaulkes committed:
fix: fixing up `setNode` garbage collection.

CMCDragonkai committed:
npm: removed un-used `uuid` dependency

tegefaulkes committed:
fix: refactored `nodeManager.setNode` garbage collection

tegefaulkes committed:
tests: proper stopping of `taskManager` in tests

tegefaulkes committed:
fix: updating task paths

`TaskPaths` should take the form of `[basePath, handlerId, ...extra]`. basePath is the `this.constructor.name` for the domain the handler is registered for.

tegefaulkes committed:
feat: task handlers are now `timedCancellable`

tegefaulkes committed:
fix: depending on return from `updateTask` to check existence

tegefaulkes committed:
fix: replacing `private` with `protected`

tegefaulkes committed:
fix: removing `Queue.ts`

tegefaulkes committed:
fix: removing `@ready` from `TaskManager.cancelTask`

tegefaulkes committed:
docs: adding description to `isConnectionError`

tegefaulkes committed:
fix: added cancellability and blocking to `nodeManager.setNode`

tegefaulkes committed:
fix: extracted `refreshBucketsDelayJitter` into nodes utils

tegefaulkes committed:
fix: fixes to errors and adding un-recoverable error handlers

Un-recoverable errors include `ErrorBinUncaughtException`, `ErrorBinUnhandledRejection` and `ErrorBinAsynchronousDeadlock`.

#414

tegefaulkes committed:
fix: moved 'syncNodeGraph` from `NodeConnectionManager` to `NodeManager`

tegefaulkes committed:
fix: `pingNode`s inside of `garbageCollectBucket` now have timeouts separate from the overall timer

other fixes have been applied.

CMCDragonkai committed:
npm: update ts-custom-error transitive dependency to 3.2.2 to fix build errors

tegefaulkes committed:
fix: cleaning up ephemeral tasks when stopping `NodeManager`

tegefaulkes committed:
fix: cleaning up errors

tegefaulkes committed:
fix: small fix to `updateRefreshBucketDelay`

tegefaulkes committed:
fix: rollback of proxy changes, was out of scope for this PR

tegefaulkes committed:
fix: small fix to `garbageCollectBucket` concurrent pinging

tegefaulkes committed:
fix: updated default timeout for `NodeConnectionManager.pingNode`

tegefaulkes committed:
fix: using Symbols for cancelling tasks

tegefaulkes committed:
fix: `TaskManager` should extend the `CreateDestroyStartStop` interface

tegefaulkes committed:
fix: test wasn't overriding key-pair generation

tegefaulkes committed:
fix: `TaskManager`'s `stopProcessing` and `stopTasks` are now properly idempotent

the @ready decorator caused them to throw if ran while `taskManager` was not running. They needed to be called during incomplete startup, so I removed the decorator.

tegefaulkes committed:
tests: slightly increasing timeouts for two tests

tegefaulkes committed:
tests: general fixes for tests failing in CI

tegefaulkes committed:
syntax: formatting change for `ExitHandlers.ts`

CMCDragonkai committed:
Merge pull request #445 from MatrixAI/feature-tasks_implementation

Integrate TaskManager into NodeGraph and Discovery

CMCDragonkai committed:
npm: updated `@matrixai/errors` to `1.1.5` so `ts-custom-error` won't cause build errors

tegefaulkes committed:
tests: extending timeouts for nat tests

tegefaulkes committed:
fix: small fixes and extending ping timeouts

tegefaulkes committed:
fix: CI handler id bug

tegefaulkes committed:
fix: extending jest timeouts for bin ping tests

tegefaulkes committed:
fix: small fix for `garbageCollectBucket`

tegefaulkes committed:
ci: adding a scratch job and test file

This allows us to run a specific test quickly via the CI during the first stage. Should allow for quicker testing.

tegefaulkes committed:
ci: fixing docker tests

tegefaulkes committed:
ci: temp disabling of most jest tests

tegefaulkes committed:
ci: disabling discovery tests

tegefaulkes committed:
wip: Checking bug

tegefaulkes committed:
wip: fix for docker test

CMCDragonkai committed:
ci: move `image` into `default` and ignore `image` for windows jobs

CMCDragonkai committed:
ci: `check:scatch` job should always be manual, and fixed syntax error
